File: benchmarks/lang/lua/fasta.lua
-- Same LCG + cumulative-table lookup as the Tzopilotl port.
local N_STEPS = 5000000
local IM = 139968
local FIM = 139968.0

local P0 = 0.27
local P1 = P0 + 0.12
local P2 = P1 + 0.12
local P3 = P2 + 0.27
local P4 = P3 + 0.02
local P5 = P3 + 0.04
local P6 = P3 + 0.06
local P7 = P3 + 0.08
local P8 = P3 + 0.10
local P9 = P3 + 0.12
local P10 = P3 + 0.14
local P11 = P3 + 0.16
local P12 = P3 + 0.18
local P13 = P3 + 0.20

local seed = 42
local c0, c1, c2, c3, c4 = 0, 0, 0, 0, 0
local c5, c6, c7, c8, c9 = 0, 0, 0, 0, 0
local c10, c11, c12, c13, c14 = 0, 0, 0, 0, 0

for _ = 1, N_STEPS do
    seed = (seed * 3877 + 29573) % IM
    local p = seed / FIM
    if     p < P0  then c0  = c0  + 1
    elseif p < P1  then c1  = c1  + 1
    elseif p < P2  then c2  = c2  + 1
    elseif p < P3  then c3  = c3  + 1
    elseif p < P4  then c4  = c4  + 1
    elseif p < P5  then c5  = c5  + 1
    elseif p < P6  then c6  = c6  + 1
    elseif p < P7  then c7  = c7  + 1
    elseif p < P8  then c8  = c8  + 1
    elseif p < P9  then c9  = c9  + 1
    elseif p < P10 then c10 = c10 + 1
    elseif p < P11 then c11 = c11 + 1
    elseif p < P12 then c12 = c12 + 1
    elseif p < P13 then c13 = c13 + 1
    else                c14 = c14 + 1 end
end

print(c0); print(c1); print(c2); print(c3); print(c4)
print(c5); print(c6); print(c7); print(c8); print(c9)
print(c10); print(c11); print(c12); print(c13); print(c14)

File: benchmarks/lang/cpp/fasta.cpp
// Same LCG + cumulative-table lookup as the Lua/Tzopilotl ports.
#include <cstdio>

static const int N_STEPS = 5000000;
static const long long IM = 139968;
static const double FIM = 139968.0;

int main() {
    const double P0  = 0.27;
    const double P1  = P0 + 0.12;
    const double P2  = P1 + 0.12;
    const double P3  = P2 + 0.27;
    const double P4  = P3 + 0.02;
    const double P5  = P3 + 0.04;
    const double P6  = P3 + 0.06;
    const double P7  = P3 + 0.08;
    const double P8  = P3 + 0.10;
    const double P9  = P3 + 0.12;
    const double P10 = P3 + 0.14;
    const double P11 = P3 + 0.16;
    const double P12 = P3 + 0.18;
    const double P13 = P3 + 0.20;

    long long seed = 42;
    long long c0=0,c1=0,c2=0,c3=0,c4=0;
    long long c5=0,c6=0,c7=0,c8=0,c9=0;
    long long c10=0,c11=0,c12=0,c13=0,c14=0;

    for (int i = 0; i < N_STEPS; ++i) {
        seed = (seed * 3877 + 29573) % IM;
        double p = seed / FIM;
        if      (p < P0 ) ++c0;
        else if (p < P1 ) ++c1;
        else if (p < P2 ) ++c2;
        else if (p < P3 ) ++c3;
        else if (p < P4 ) ++c4;
        else if (p < P5 ) ++c5;
        else if (p < P6 ) ++c6;
        else if (p < P7 ) ++c7;
        else if (p < P8 ) ++c8;
        else if (p < P9 ) ++c9;
        else if (p < P10) ++c10;
        else if (p < P11) ++c11;
        else if (p < P12) ++c12;
        else if (p < P13) ++c13;
        else              ++c14;
    }

    std::printf("%lld\n%lld\n%lld\n%lld\n%lld\n", c0, c1, c2, c3, c4);
    std::printf("%lld\n%lld\n%lld\n%lld\n%lld\n", c5, c6, c7, c8, c9);
    std::printf("%lld\n%lld\n%lld\n%lld\n%lld\n", c10, c11, c12, c13, c14);
}
